### PR TITLE
fix(subparsers): required

### PR DESCRIPTION
**Pull request by scintillating7**
The usual way to do argparsing is something like:
```
# create the top-level parser
parser = argparse.ArgumentParser()
subparsers = parser.add_subparsers()

# create the parser for the "foo" command
parser_foo = subparsers.add_parser('foo')
parser_foo.add_argument('-x', type=int)

args = parser.parse_args('foo 1 -x 2'.split())
args.func(args)
```

Python3 [argparse.subparsers](https://docs.python.org/3/library/argparse.html#sub-commands) have their `required` flag set to `False` by default. The old 2.7 `parse_args` codepath would throw an error if we got [this far](https://github.com/python/cpython/blob/2.7/Lib/argparse.py#L1954) without finding any arguments, but the newer version [lacks that check](https://github.com/python/cpython/blob/master/Lib/argparse.py#L2071). That means that given a subparser not marked as `required==True`, given no arguments, the arg parsing will pass and return `args=args.Namespace({'commands': None})`. 

TL;DR: if you enter just a command with no args, you will end up getting a somewhat opaque `AttributeError: 'Namespace' object has no attribute 'func'` error in your last line instead of invoking the foo handler.

This is a pretty well known [bug/feature](https://stackoverflow.com/questions/22990977/why-does-this-argparse-code-behave-differently-between-python-2-and-3/22994500#22994500) and it seems marking the subparser as required is the easiest fix.

Also cleans up some dead code.

Addresses https://jira.opensciencedatacloud.org/browse/SV-1648

### CODE REVIEW

**Comment by philtom-ctds** [comment]
> The usual way to do argparsing is something like:
> 
> ```
> # create the top-level parser
> parser = argparse.ArgumentParser()
> subparsers = parser.add_subparsers()
> 
> # create the parser for the "foo" command
> parser_foo = subparsers.add_parser('foo')
> parser_foo.add_argument('-x', type=int)
> 
> args = parser.parse_args('foo 1 -x 2'.split())
> args.func(args)
> ```
> 
> Python3 [argparse.subparsers](https://docs.python.org/3/library/argparse.html#sub-commands) have their `required` flag set to `False` by default. The old 2.7 `parse_args` codepath would throw an error if we got [this far](https://github.com/python/cpython/blob/2.7/Lib/argparse.py#L1954) without finding any arguments, but the newer version [lacks that check](https://github.com/python/cpython/blob/master/Lib/argparse.py#L2071). That means that given a subparser not marked as `required==True`, given no arguments, the arg parsing will pass and return `args=args.Namespace({'commands': None})`.
> 
> TL;DR: if you enter just a command with no args, you will end up getting a somewhat opaque `AttributeError: 'Namespace' object has no attribute 'func'` error in your last line instead of invoking the foo handler.
> 
> This is a pretty well known [bug/feature](https://stackoverflow.com/questions/22990977/why-does-this-argparse-code-behave-differently-between-python-2-and-3/22994500#22994500) and it seems marking the subparser as required is the easiest fix.
> 
> Also cleans up some dead code.
> 
> Addresses https://jira.opensciencedatacloud.org/browse/SV-1648

This would make an awesome commit message.  It's explains the "why", while the diff explains the "what".

**Review by anmaxvl — APPROVED** [review]

**Comment by scintillating7** [comment]
- required - Whether or not a subcommand must be provided, by default False (added in 3.7)

FWIW, you can pass `required` into the constructor for a subparser as of Python3.7. But we're on 3.5/6, so adding as attribute after the fact is the way to go.

File: bin/gdc-client [hunk 1]
```diff
@@ -76,6 +76,7 @@ if __name__ == '__main__':
         dest='command',
         help='for more information, specify -h after a command',
     )
+    subparsers.required = True
```
Comment (ProfOak) [inline bin/gdc-client:79]: Can we add this as a method parameter instead?

https://docs.python.org/3/library/argparse.html#argparse.ArgumentParser.add_subparsers

File: gdc_client/settings/parser.py [hunk 2]
```diff
@@ -33,6 +33,7 @@ def resolve(config_file, args):
 def config(parser, config_file=None):
     parser.add_argument('--config', help=HELP, metavar='FILE')
     choices = parser.add_subparsers(title='Settings to display', dest='section')
+    choices.required = True
```
Comment (ProfOak) [inline gdc_client/settings/parser.py:36]: Same here.

**Review by ProfOak — COMMENTED** [review]
~I had a couple of minor style concerns.~
:sweat_smile:

**Review by ProfOak — APPROVED** [review]

**Review by philtom-ctds — APPROVED** [review]
